src/lib.rs

```diff
@@ -234,6 +234,38 @@ impl GetAdjacencyMatrix for PyDAG {
     }
 }
 
+// Rust side only PyDAG methods
+impl PyDAG {
+    fn _add_edge(
+        &mut self,
+        p_index: NodeIndex,
+        c_index: NodeIndex,
+        edge: PyObject,
+    ) -> PyResult<usize> {
+        if self.check_cycle {
+            let should_check_for_cycle =
+                must_check_for_cycle(self, p_index, c_index);
+            let state = Some(&mut self.cycle_state);
+            if should_check_for_cycle
+                && algo::has_path_connecting(
+                    &self.graph,
+                    c_index,
+                    p_index,
+                    state,
+                )
+            {
+                Err(DAGWouldCycle::py_err("Adding an edge would cycle"))
+            } else {
+                let edge = self.graph.add_edge(p_index, c_index, edge);
+                Ok(edge.index())
+            }
+        } else {
+            let edge = self.graph.add_edge(p_index, c_index, edge);
+            Ok(edge.index())
+        }
```

Comment:
Could these be combined by merging the logic again once an error isn't raised?

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/82/commits/98d316eeaff8db546fbec33f80370279de193cd9